Elastic failure handling

krfricke opened the pull request:
Introduces elastic failure handling. Alive actors are not recreated on error, but stick around and don't have to load data again. We can also choose to not restart failed actors, continuing training with fewer actors.

Depends on #21.

richardliaw commented:
@amogkam can you review first?

also, the test is timing out. I think we should also put a time limit (10 minutes or something) on GHA :)

krfricke commented:
Yep, I'll try to fix the test today and will introduce the timeout

krfricke commented:
It seems some resources are not cleaned up properly. I noticed that with the queue actor before, maybe it affects the event driver, too. I'll look into this.

krfricke commented:
This should be ready for review @amogkam

amogkam commented:
Thanks I'll review it today

amogkam commented on xgboost_ray/main.py:
```diff
@@ -23,22 +24,28 @@
+    pass
+
+
+class RayXGBoostTrainingStopped(RuntimeError):
```
Can you add a comment describing what errors are and the difference between them

amogkam commented on xgboost_ray/tests/test_fault_tolerance.py:
```diff
@@ -87,25 +113,98 @@ def setUp(self):
+                max_actor_restarts=1,
+                num_actors=2,
+                elastic_training=True,
+                max_failed_actors=1),
```
This should raise an error if `max_failed_actors` is set to 0 right? Is it worth adding another test for that case?

krfricke replied:
Added `testTrainingStopElastic`

amogkam reviewed (commented):
This is a really cool feature and looks good to me overall! The main thing is that there are a lot of different concurrency components (multiprocessing, threading, asyncio.Event, rabit_tracker). I know you have some of this in the Google doc already but it would really help to add some comments here about what each component's responsibilities are and how they interact with each other.

amogkam commented on xgboost_ray/main.py:
```diff
@@ -157,6 +168,11 @@ class RayParams:
+        max_failed_actors (int): If `elastic_training` is True, this
+            specifies the maximum number of failed actors with which
+            we still continue training.
         max_actor_restarts (int): Number of retries when Ray actors fail.
```
Ok it looks like max_actor_restarts is still used for when there are training failures, but the actors themselves are still alive, is that correct?

krfricke replied:
Yep, you're right this is confusing. I think it makes sense to ignore this when `elastic_training` is enabled. I added a note to the docs and implemented the logic

krfricke commented on xgboost_ray/util.py:
```diff
@@ -0,0 +1,45 @@
+
+
+class Queue(RayQueue):
+    def shutdown(self):
```
Yeah I think that makes sense. I'll file a PR later (and also add the Event class)

krfricke commented on xgboost_ray/main.py:
```diff
@@ -61,11 +68,15 @@ def _start_rabit_tracker(num_workers: int):
-    thread = Thread(target=rabit_tracker.join)
-    thread.daemon = True
-    thread.start()
+    process = multiprocessing.Process(target=rabit_tracker.join)
```
So the Rabit tracker is the "driver"/head process of the Rabit allreduce implementation. All Rabit actors register with the tracker.
We're running the tracker on the driver and not in a Ray process or actor. After pausing or killing actors it can lead to problems if we re-use the same tracker instance, so we have to create a new one. In order to properly clean up the old tracker, we have to terminate the process somehow. Since python threads don't support killing, we're starting a process here that is hence killable.

amogkam replied:
Got it that makes sense! Do you think you can add this as a comment in the code?

richardliaw replied:
yep, had the same question too! A comment in the code would be much appreciated.

krfricke commented on xgboost_ray/main.py:
```diff
@@ -219,21 +237,43 @@ def __init__(self,
+        """Get process PID. Used for checking if still alive"""
+        return os.getpid()
+
+    def _stop_training_callback(self):
```
It stops training if the event is set. `asyncio.Event` is just the internal structure the Event ray actor uses (just like the Queue ray actor uses `asyncio.Queue`. However I don't think we need this callback anymore since callbacks aren't processed anymore once a local training actor died. I'll remove it

krfricke commented on xgboost_ray/main.py:
```diff
@@ -219,21 +237,43 @@ def __init__(self,
+        this = self
+
+        class _StopCheckpointCallback(TrainingCallback):
+            if this._stop_event and this._stop_event.is_set():
```
Yep, thanks for the catch. Removed the whole callback (see above).

krfricke commented on xgboost_ray/main.py:
```diff
@@ -419,39 +524,70 @@ def _train(params: Dict,
     # Create remote actors
-    actors = [
-        _create_actor(
+    def create_actor(*args, **kwargs):
```
Haha good catch. This is a leftover from the hackathon where we used failure callbacks. Removed!

krfricke commented on xgboost_ray/main.py:
```diff
@@ -295,21 +340,44 @@ def train(self, rabit_args: List[str], params: Dict[str, Any],
+                # Silent fail, will be raised as RayXGBoostTrainingStopped
+                return
+
+        thread = threading.Thread(target=_train)
```
We need the thread to be able to react to the stop event. I'll add a comment.

krfricke commented on xgboost_ray/main.py:
```diff
@@ -295,21 +340,44 @@ def train(self, rabit_args: List[str], params: Dict[str, Any],
+        thread.start()
+        while thread.is_alive():
+            thread.join(timeout=0)
+            if self._stop_event.is_set():
```
In case of failure of actor A, this will stop training of the other actors (B, C, etc). If we didn't terminate the train() function (which we couldn't if we didn't use threads) the actors would be stuck and we couldn't start a new train() process on them

amogkam replied:
Got it, makes sense! Do you think you can add this ^ as a comment in the code?

krfricke commented on xgboost_ray/main.py:
```diff
@@ -359,28 +427,61 @@ def _handle_queue(queue: Queue, checkpoint: _Checkpoint,
 
 
-def _shutdown(remote_workers: List[ActorHandle],
+def _get_actor_alive_status(actors: List[ActorHandle],
```
Done

krfricke commented on xgboost_ray/main.py:
```diff
@@ -588,22 +741,43 @@ def train(params: Dict,
+                stop_event.shutdown()
+                time.sleep(5)
+                queue = Queue()
+                stop_event = Event()
```
Theoretically we could keep them, practically I saw queue actors sticking around after program termination, so I recreate them to have a clean communication slate between driver and actors. Is there any major disadvantage doing so? Scheduling should be low overhead, especially compared to training times.

richardliaw replied:
Killing is probably OK for now, we can adjust it after some initial feedback.

krfricke commented on xgboost_ray/tests/test_fault_tolerance.py:
```diff
@@ -20,6 +20,32 @@ def tree_obj(bst: xgb.Booster):
+                    epoch == fail_iteration and \
+                    not os.path.exists(die_lock_file):
+                # Only die once
+                if os.path.exists(die_lock_file):
```
Yep! Thanks!

krfricke commented on xgboost_ray/main.py:
```diff
@@ -359,28 +427,61 @@ def _handle_queue(queue: Queue, checkpoint: _Checkpoint,
+                ray.get(actor.__ray_terminate__.remote())
+            except RayActorError:
+                ray.kill(actor)
+        # actors[i] = None
```
Forgot about that one - actually it should be uncommented. I needed it for testing but went with a mock side effect instead to track alive actors.

krfricke commented:
Thanks a bunch for the thorough review. I addressed all your comments and tried to answer your questions!

richardliaw commented:
I’ll review this PR later tonight!

On Fri, Dec 11, 2020 at 9:11 AM Amog Kamsetty <notifications@github.com>
wrote:

> *@amogkam* commented on this pull request.
> ------------------------------
>
> In xgboost_ray/main.py
> <https://github.com/ray-project/xgboost_ray/pull/23#discussion_r541097819>
> :
>
> > +                        **kwargs)
> +                    result_dict.update({
> +                        "bst": bst,
> +                        "evals_result": evals_result,
> +                        "train_n": self._local_n
> +                    })
> +            except XGBoostError:
> +                # Silent fail, will be raised as RayXGBoostTrainingStopped
> +                return
> +
> +        thread = threading.Thread(target=_train)
> +        thread.daemon = True
> +        thread.start()
> +        while thread.is_alive():
> +            thread.join(timeout=0)
> +            if self._stop_event.is_set():
>
> Got it, makes sense! Do you think you can add this ^ as a comment in the
> code?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/xgboost_ray/pull/23#discussion_r541097819>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCRZZI42L7RWJA3DD3IHSTSUJHCZANCNFSM4UQURGRA>
> .
>

richardliaw commented on xgboost_ray/main.py:
```diff
@@ -359,28 +424,64 @@ def _handle_queue(queue: Queue, checkpoint: _Checkpoint,
-            logger.warning("Failed to shutdown gracefully, forcing a "
-                           "shutdown.")
-            _shutdown(remote_workers, force=True)
+    for i in range(len(actors)):
+        actor = actors[i]
```
for actor in actors?

krfricke replied:
Since we're overwriting the array item here:
```
        actors[i] = None
``` we need the index anyway, so I'll probably keep it as is

richardliaw commented on xgboost_ray/main.py:
```diff
@@ -588,22 +737,45 @@ def train(params: Dict,
             else:
                 raise RuntimeError(
-                    "A Ray actor died during training and the maximum number "
-                    "of retries ({}) is exhausted.")
+                    f"A Ray actor died during training and the maximum number "
+                    f"of retries ({max_actor_restarts}) is exhausted.")
```
raise from...

richardliaw commented on xgboost_ray/main.py:
```diff
@@ -419,39 +524,67 @@ def _train(params: Dict,
+    # Maybe we got a new queue actor
+    wait_queue = [
+        actor.set_queue.remote(_queue) for actor in _actors
+        if actor is not None
```
can there actually be `actor is None` here?

krfricke replied:
Yes! After an actor failed its entry in `_actors` is set to None in the `train()` function. `_train` is then called again, so entries can be None

richardliaw replied:
Hmm, don't all the None entries get filled in L586 to 603?

krfricke replied:
Only those that are in `_failed_actor_ranks` - and this set is empty if we continue with fewer actors. See previous comment below

richardliaw commented on xgboost_ray/main.py:
```diff
@@ -419,39 +524,67 @@ def _train(params: Dict,
-        _shutdown(actors, queue, force=True)
-        raise
+    except Exception as exc:
+        _stop_event.set()
+        _get_actor_alive_status(_actors, handle_actor_failure)
+        raise RayActorError from exc
```
maybe we can do this in a separate PR.

krfricke replied:
`_train()` is supposed to be a single multi-actor run. `train()` is supposed to be the full run, which triggers one or more invocations of `_train()` - i.e. fault tolerance handling should be done in `train()`. `_train()` should just make sure to clean up its references to actors after failure. Let's chat about this though to see how I can improve the logic here.

richardliaw replied:
Sounds good; we can do that later!

richardliaw reviewed (commented):
This looks pretty good overall! A couple comments

1. The architecture/design could be better commented in the code. I left some comments at places that required a bit more digging.
2. It'd be great to document this failure model in `train(`. 
3. We will want to do a bit of refactoring to consolidate failure handling responsibilities in core functions, but I don't think that's a blocker for this PR.

Please ping when you've addressed the remaining comments!

richardliaw commented:
Actually, 1 question I have is -- 

when you have 4 actors on 4 nodes, and 1 of the nodes die, where do you check the number of cluster resources to resume with only 3 workers?

krfricke commented:
Since we re-use the existing actors, we currently don't check the cluster resources at all - we just invoke the local `train()` function on the remaining actors.

Thanks a  lot for the review by the way! I addressed the comments and just added a bunch of in-code documentation for better understanding of the code.

richardliaw commented on xgboost_ray/main.py:
```diff
@@ -419,100 +570,167 @@ def _train(params: Dict,
+    # will be all actors. In future invocations, this may be less than
+    # the num_actors setting, depending on the failure mode.
+    newly_created = 0
+    for i in list(_failed_actor_ranks):
+        if _actors[i] is not None:
+            raise RuntimeError(
+                f"Trying to create actor with rank {i}, but it already "
+                f"exists.")
+        actor = _create_actor(
```
Question - from what I understand, failed actors will have their rank added to `_failed_actor_ranks`. When `_train` is called, don't these actors get recreated?

I'm just specifically looking for the situation where a node dies and new actor creation  step is skipped

krfricke replied:
That's in line 848:
```
                 start_actor_ranks = set() 
```
here the set is cleared (I guess we could just `.clear()` it instead). With it being empty, no actors are restarted

richardliaw replied:
Ah, got it.

richardliaw commented:
Unfortunately, my cluster yaml (and snapshot) no longer works. One last comment/question about actor recovery -- I keep re-reading the recovery code and it seems like the actors are always restarted. Can you help explain the purpose/code flow of `_failed_actor_ranks`?

krfricke commented:
I'll run a couple of tests on a cluster later and post the results here.

krfricke commented:
Re: `_failed_actor_ranks`: This set collects the ranks of failed actors. These are restarted on a call to `_train`. Note that we have the same information in the `_actors` list, as we could just filter for `None` values - however, the difference here is that we want to be able to deliberately alter the `_failed_actor_ranks` in order to re-start only specific nodes.
For instance, for elastic training we currently clear the set, so no actors will be restarted. In future failure handling modes we could only start actors that have resources available. With this object we can thus specify exactly which ranks we want to start and which not.

In the `train()` function the set is called `start_actor_ranks`, as it semantically describes which ranks will be started by `_train()`. In `_train()` though it is called `_failed_actor_ranks` as, after starting actors, it collects the ranks of failed actors. Would it be more clear if we renamed `_failed_actor_ranks` to `_start_actor_ranks` in `_train()` as well?

I can also add more comments to this - e.g. annotate the internal variables passed to `_train()`.

krfricke commented:
Works fine in my tests. Here is the results log for killing 2 nodes:

```
base) root@ip-172-31-18-125:/release_tests# python benchmark_cpu_gpu.py 10 100 200
2020-12-14 10:42:37,239 INFO worker.py:651 -- Connecting to existing Ray cluster at address: 172.31.18.125:6379
2020-12-14 10:42:37,992 INFO main.py:620 -- [RayXGBoost] Created 10 new actors (10 total actors).
2020-12-14 10:42:40,938 INFO main.py:638 -- [RayXGBoost] Starting XGBoost training.
(pid=4107, ip=172.31.20.197) [10:42:40] task [xgboost.ray]:140004211236752 got new rank 2
(pid=4038, ip=172.31.25.131) [10:42:40] task [xgboost.ray]:140546107764624 got new rank 5
(pid=4005, ip=172.31.24.220) [10:42:40] task [xgboost.ray]:140103861936592 got new rank 4
(pid=4006, ip=172.31.29.248) [10:42:40] task [xgboost.ray]:140411510034704 got new rank 7
(pid=4037, ip=172.31.17.26) [10:42:40] task [xgboost.ray]:139927338557264 got new rank 0
(pid=3993, ip=172.31.20.30) [10:42:40] task [xgboost.ray]:140255472527504 got new rank 3
(pid=4074, ip=172.31.26.109) [10:42:40] task [xgboost.ray]:140038723952528 got new rank 6
(pid=3939, ip=172.31.30.64) [10:42:40] task [xgboost.ray]:140290278254992 got new rank 9
(pid=4107, ip=172.31.30.251) [10:42:40] task [xgboost.ray]:140133743529872 got new rank 8
(pid=4011, ip=172.31.19.132) [10:42:40] task [xgboost.ray]:140660999946128 got new rank 1
E1214 10:42:53.648838  3040  3055 task_manager.cc:323] Task failed: IOError: 2: user code caused exit: Type=ACTOR_TASK, Language=PYTHON, Resources: {}
, function_descriptor={type=PythonFunctionDescriptor, module_name=xgboost_ray.main, class_name=RayXGBoostActor, function_name=train, function_hash=},
task_id=019554c512c2b9b51bd49a5609000000, task_name=RayXGBoostActor.train(), job_id=09000000, num_args=10, num_returns=2, actor_task_spec={actor_id=1b
d49a5609000000, actor_caller_id=ffffffffffffffffffffffff09000000, actor_counter=4}
E1214 10:42:53.650756  3040  3055 task_manager.cc:323] Task failed: IOError: cancelling all pending tasks of dead actor: Type=ACTOR_TASK, Language=PYT
HON, Resources: {}, function_descriptor={type=PythonFunctionDescriptor, module_name=xgboost_ray.main, class_name=RayXGBoostActor, function_name=pid, f
unction_hash=}, task_id=b9fed017052010151bd49a5609000000, task_name=RayXGBoostActor.pid(), job_id=09000000, num_args=0, num_returns=2, actor_task_spec
={actor_id=1bd49a5609000000, actor_caller_id=ffffffffffffffffffffffff09000000, actor_counter=5}
2020-12-14 10:42:53,729 INFO main.py:489 -- Actor status: 9 alive, 1 dead (10 total)
2020-12-14 10:42:53,729 WARNING main.py:851 -- A Ray actor died during training. Trying to restart and continue training from last checkpoint (restart
 1 of 2). This will use 9 existing actors and start 0 new actors.Sleeping for 10 seconds for cleanup.
2020-12-14 10:43:03,750 INFO main.py:620 -- [RayXGBoost] Created 0 new actors (9 total actors).
2020-12-14 10:43:03,758 INFO main.py:638 -- [RayXGBoost] Starting XGBoost training.
(pid=4107, ip=172.31.20.197) [10:43:03] task [xgboost.ray]:140004211236752 got new rank 2
(pid=4038, ip=172.31.25.131) [10:43:03] task [xgboost.ray]:140546107764624 got new rank 5
(pid=4011, ip=172.31.19.132) [10:43:03] task [xgboost.ray]:140660999946128 got new rank 1
(pid=4037, ip=172.31.17.26) [10:43:03] task [xgboost.ray]:139927338557264 got new rank 0
(pid=4006, ip=172.31.29.248) [10:43:03] task [xgboost.ray]:140411510034704 got new rank 6
(pid=3993, ip=172.31.20.30) [10:43:03] task [xgboost.ray]:140255472527504 got new rank 3
(pid=4005, ip=172.31.24.220) [10:43:03] task [xgboost.ray]:140103861936592 got new rank 4
(pid=4107, ip=172.31.30.251) [10:43:03] task [xgboost.ray]:140133743529872 got new rank 7
(pid=3939, ip=172.31.30.64) [10:43:03] task [xgboost.ray]:140290278254992 got new rank 8
E1214 10:43:15.269685  3040  3055 task_manager.cc:323] Task failed: IOError: 2: user code caused exit: Type=ACTOR_TASK, Language=PYTHON, Resources: {}
, function_descriptor={type=PythonFunctionDescriptor, module_name=xgboost_ray.main, class_name=RayXGBoostActor, function_name=train, function_hash=},
task_id=aff64b4aedc094a1f90e593909000000, task_name=RayXGBoostActor.train(), job_id=09000000, num_args=12, num_returns=2, actor_task_spec={actor_id=f9
0e593909000000, actor_caller_id=ffffffffffffffffffffffff09000000, actor_counter=10}
E1214 10:43:15.271399  3040  3055 task_manager.cc:323] Task failed: IOError: cancelling all pending tasks of dead actor: Type=ACTOR_TASK, Language=PYT
HON, Resources: {}, function_descriptor={type=PythonFunctionDescriptor, module_name=xgboost_ray.main, class_name=RayXGBoostActor, function_name=pid, f
unction_hash=}, task_id=b55dd8e99d574a31f90e593909000000, task_name=RayXGBoostActor.pid(), job_id=09000000, num_args=0, num_returns=2, actor_task_spec
={actor_id=f90e593909000000, actor_caller_id=ffffffffffffffffffffffff09000000, actor_counter=11}
2020-12-14 10:43:15,367 INFO main.py:489 -- Actor status: 8 alive, 2 dead (10 total)
2020-12-14 10:43:15,367 WARNING main.py:851 -- A Ray actor died during training. Trying to restart and continue training from last checkpoint (restart
 2 of 2). This will use 8 existing actors and start 0 new actors.Sleeping for 10 seconds for cleanup.
2020-12-14 10:43:25,388 INFO main.py:620 -- [RayXGBoost] Created 0 new actors (8 total actors).
2020-12-14 10:43:25,394 INFO main.py:638 -- [RayXGBoost] Starting XGBoost training.
(pid=4006, ip=172.31.29.248) [10:43:25] task [xgboost.ray]:140411510034704 got new rank 5
(pid=3993, ip=172.31.20.30) [10:43:25] task [xgboost.ray]:140255472527504 got new rank 2
(pid=4005, ip=172.31.24.220) [10:43:25] task [xgboost.ray]:140103861936592 got new rank 3
(pid=3939, ip=172.31.30.64) [10:43:25] task [xgboost.ray]:140290278254992 got new rank 7
(pid=4107, ip=172.31.30.251) [10:43:25] task [xgboost.ray]:140133743529872 got new rank 6
(pid=4037, ip=172.31.17.26) [10:43:25] task [xgboost.ray]:139927338557264 got new rank 0
(pid=4011, ip=172.31.19.132) [10:43:25] task [xgboost.ray]:140660999946128 got new rank 1
(pid=4038, ip=172.31.25.131) [10:43:25] task [xgboost.ray]:140546107764624 got new rank 4
(pid=4074, ip=172.31.26.109) [10:42:40] task [xgboost.ray]:140038723952528 got new rank 6
(pid=4074, ip=172.31.26.109) 2020-12-14 10:42:53,643    ERROR worker.py:384 -- SystemExit was raised from the worker
(pid=4074, ip=172.31.26.109) Traceback (most recent call last):
(pid=4074, ip=172.31.26.109)   File "python/ray/_raylet.pyx", line 551, in ray._raylet.task_execution_handler
(pid=4074, ip=172.31.26.109)   File "python/ray/_raylet.pyx", line 438, in ray._raylet.execute_task
(pid=4074, ip=172.31.26.109)   File "python/ray/_raylet.pyx", line 477, in ray._raylet.execute_task
(pid=4074, ip=172.31.26.109)   File "python/ray/_raylet.pyx", line 481, in ray._raylet.execute_task
(pid=4074, ip=172.31.26.109)   File "python/ray/_raylet.pyx", line 482, in ray._raylet.execute_task
(pid=4074, ip=172.31.26.109)   File "python/ray/_raylet.pyx", line 436, in ray._raylet.execute_task.function_executor
(pid=4074, ip=172.31.26.109)   File "/root/anaconda3/lib/python3.7/site-packages/ray/function_manager.py", line 553, in actor_method_executor
(pid=4074, ip=172.31.26.109)     return method(actor, *args, **kwargs)
(pid=4074, ip=172.31.26.109)   File "/root/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 390, in train
(pid=4074, ip=172.31.26.109)     time.sleep(0.1)
(pid=4074, ip=172.31.26.109)   File "/root/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 381, in sigterm_handler
(pid=4074, ip=172.31.26.109)     sys.exit(1)
(pid=4074, ip=172.31.26.109) SystemExit: 1
(pid=4107, ip=172.31.20.197) [10:42:40] task [xgboost.ray]:140004211236752 got new rank 2
(pid=4107, ip=172.31.20.197) [10:43:03] task [xgboost.ray]:140004211236752 got new rank 2
(pid=4107, ip=172.31.20.197) 2020-12-14 10:43:15,245    ERROR worker.py:384 -- SystemExit was raised from the worker
(pid=4107, ip=172.31.20.197) Traceback (most recent call last):
(pid=4107, ip=172.31.20.197)   File "python/ray/_raylet.pyx", line 551, in ray._raylet.task_execution_handler
(pid=4107, ip=172.31.20.197)   File "python/ray/_raylet.pyx", line 438, in ray._raylet.execute_task
(pid=4107, ip=172.31.20.197)   File "python/ray/_raylet.pyx", line 477, in ray._raylet.execute_task
(pid=4107, ip=172.31.20.197)   File "python/ray/_raylet.pyx", line 481, in ray._raylet.execute_task
(pid=4107, ip=172.31.20.197)   File "python/ray/_raylet.pyx", line 482, in ray._raylet.execute_task
(pid=4107, ip=172.31.20.197)   File "python/ray/_raylet.pyx", line 436, in ray._raylet.execute_task.function_executor
(pid=4107, ip=172.31.20.197)   File "/root/anaconda3/lib/python3.7/site-packages/ray/function_manager.py", line 553, in actor_method_executor
(pid=4107, ip=172.31.20.197)     return method(actor, *args, **kwargs)
(pid=4107, ip=172.31.20.197)   File "/root/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 390, in train
(pid=4107, ip=172.31.20.197)     time.sleep(0.1)
(pid=4107, ip=172.31.20.197)   File "/root/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 381, in sigterm_handler
(pid=4107, ip=172.31.20.197)     sys.exit(1)
(pid=4107, ip=172.31.20.197) SystemExit: 1
2020-12-14 10:45:12,824 INFO main.py:731 -- [RayXGBoost] Finished XGBoost training on training data with total N=160,000,000.
TRAIN TIME TAKEN: 155.57 seconds
Final training error: 0.4951
TOTAL TIME TAKEN: 155.58 seconds (0.03 for init)
(base) root@ip-172-31-18-125:/release_tests#
```

Note that the last two errors are sent to the driver some time after the nodes died.

I'll push some cosmetic changes, but other than I think we should be ready to merge.

richardliaw commented:
nice!
